src/List/__tests__/item.spec.js

```diff
@@ -0,0 +1,75 @@
+import { createElement } from 'react';
+import { configure, shallow } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+import { Item } from '../index.js';
+
+configure({ adapter: new Adapter() });
+
+const classes = {
+    root: 'abc'
+};
+
+test('renders a div by default', () => {
+    const props = { classes, item: 'a' };
+    const wrapper = shallow(<Item {...props} />).dive();
+
+    expect(wrapper.type()).toEqual('div');
+    expect(wrapper.prop('className')).toEqual(classes.root);
+});
+
+test('renders a provided tagname', () => {
+    const props = { classes, item: 'a', render: 'p' };
+    const wrapper = shallow(<Item {...props} />).dive();
+
+    expect(wrapper.type()).toEqual('p');
+    expect(wrapper.prop('className')).toEqual(classes.root);
+});
+
+test('renders a provided component', () => {
+    const Span = () => <span />;
+    const props = { item: 'a', render: Span };
+    const wrapper = shallow(<Item {...props} />);
+
+    expect(wrapper.type()).toEqual(Span);
+    expect(wrapper.dive().type()).toEqual('span');
+});
+
+test('passes only rest props to basic `render`', () => {
+    const props = { classes, item: 'a', render: 'p' };
+    const wrapper = shallow(<Item {...props} data-id="b" />).dive();
+
+    expect(wrapper.props()).toHaveProperty('data-id');
+    expect(wrapper.props()).not.toHaveProperty('classes');
+    expect(wrapper.props()).not.toHaveProperty('hasFocus');
+    expect(wrapper.props()).not.toHaveProperty('isSelected');
+    expect(wrapper.props()).not.toHaveProperty('item');
+    expect(wrapper.props()).not.toHaveProperty('render');
+});
+
+test('passes custom and rest props to composite `render`', () => {
+    const Span = () => <span />;
+    const props = { classes, item: 'a', render: Span };
+    const wrapper = shallow(<Item {...props} data-id="b" />);
+
+    expect(wrapper.props()).toHaveProperty('data-id');
+    expect(wrapper.props()).toHaveProperty('classes');
+    expect(wrapper.props()).toHaveProperty('hasFocus');
+    expect(wrapper.props()).toHaveProperty('isSelected');
+    expect(wrapper.props()).toHaveProperty('item');
+    expect(wrapper.props()).not.toHaveProperty('render');
+});
+
+test('passes `item` string as `children`', () => {
+    const props = { item: 1, render: 'p' };
+    const wrapper = shallow(<Item {...props} />).dive();
+
+    expect(wrapper.text()).toEqual('1');
+});
+
+test('renders `item` string regardless of type', () => {
+    const props = { item: { id: 1 }, render: 'p' };
+    const wrapper = shallow(<Item {...props} />).dive();
+
+    expect(wrapper.text()).toEqual('[object Object]');
```

Comment:
This is going to be a source of bugs. Right now, `item` is set to an `any` proptype, and we're explicitly coercing it to a `string` in the render method.

We can help people catch errors earlier by making that a `proptype` of `string`, and have the user coerce the value themselves if they have a non-string value they're providing.

Reply:
We can't expect `item` to be a `string`. It can be a string or an object; in most cases it'll be an object of arbitrary shape.

If `item` is an object, since we can't predict its shape, you're expected to also provide a function or class to `render` that knows how to parse the `item` object. If `item` is a string, though, you can just provide a tagname to `render` and we'll render it for you.

Instead of just coercing `item` to a string, what if we put a condition in there to only render `item` if it's already a string?